runtime/src/threads/pthreads/threads-pthreads.c

```diff
@@ -81,6 +83,102 @@ static void            (*saved_threadEndFn)(void);
 static void*           initial_pthread_func(void*);
 static void*           pthread_func(void*);
 
+/* This conditional is necessary because on some operating systems
+ * (like OSX) the cstdlib malloc doesn't use nmap for the allocation
+ * and can't be mprotect-ed. In that case, we use the default pthread
+ * stack
+ */
+#ifndef CHPL_USING_CSTDLIB_MALLOC
```

Comment:
Ah, I'd forgotten that detail. Currently this test distinguishes between allowing pthreads to allocate its own stacks, versus allocating those stacks from our own Chapel heap (in which case we always create the guard page).

I could instead always allocate the stacks ourselves (because that has benefits such as tracking) and enable/disable creating the guard page based on the architecture or existence of a registered heap. So basically we would end up making most of the new code in this file unconditional, and only protecting the code that mprotect()s and un-mprotect()s the guard pages depend on the architecture and/or the presence of a registered heap.  If this were so, I would not mind if we checked CHPL_USING_CSTDLIB_MALLOC, though only when #if defined(__APPLE__). I would also want to compare system and heap page sizes, though.

Michael, how do you feel? I'd like to do all this as unconditionally as we can (on principle, not because I think there's much performance effect). Any problem with just always getting the stacks from the Chapel allocator, and doing guard pages when we can?  Put another way, when the heap is on hugepages, would you rather have unregistered stacks with guard pages, or registered stacks without guard pages? I think I'm mildly in favor of the latter, but I could be swayed.  (Note that we could also support both and choose between them at execution time based on the value of CHPL_STACK_CHECKS; see tasks-qthreads.c.)